File: coding_int/chapter_1/1_5_OneAway.cpp
#include<iostream>
#include<unordered_map>
#include<string>
#include<cmath>

bool oneAway(std::string a, std::string b){

    std::unordered_map<char, int> a_map;
    std::unordered_map<char, int> b_map;

    int diff = a.length() - b.length();

    int length_diff = std::abs(diff);
    if (length_diff > 1) {
        return false;
    }
    //std::cout << "same length" << std::endl;

    for (std::string::size_type i = 0; i < a.size(); ++i) {
        ++a_map[a[i]];
    }

    for (std::string::size_type i = 0; i < b.size(); ++i) {
        ++b_map[b[i]];
    }

    int diff_count = 0;

    for (auto it : a_map) {
        auto b_find = b_map.find(it.first);
        //std::cout << "into loop" << std::endl;
        if (it.second > 1) {
            //std::cout << "path 1 " << std::endl;
            if (b_find == b_map.end()) {
                return false;
            } else {
                int bcount = b_find->second;
                int count_diff = it.second - bcount;
                if (count_diff > 1) {
                    return false;
                } else if (count_diff == 1) {
                    ++diff_count;
                }
            }
        }

        if (it.second == 1) {
            //std::cout << "path 2 " << std::endl;
            if (b_find == b_map.end()) {
                ++diff_count;
            } else {
                int bcount = b_find->second;
                int count_diff = it.second - bcount;
                if (count_diff > 1) {
                    return false;
                } else if (count_diff == 1) {
                    ++diff_count;
                }
            }
        }

        if (diff_count > 1) {
            return false;
        }  
    }

    return true;
}

int main() {

    std::string a = "pale";
    std::string b = "ple";

    std::string c = "pales";
    std::string d = "pale";
    
    std::string e = "pale";
    std::string f = "bale";

    std::string g = "pale";
    std::string h = "bake";

    std::cout << "string a " << a << " string b " << b << " result " << oneAway(a,b) << std::endl;
    std::cout << "string a " << c << " string b " << d << " result " << oneAway(c,d) << std::endl;
    std::cout << "string a " << e << " string b " << f << " result " << oneAway(e,f) << std::endl;
    std::cout << "string a " << g << " string b " << h << " result " << oneAway(g,h) << std::endl;
    
}

File: coding_int/chapter_1/1_5_OneAway.py
from collections import Counter

def oneAway(a,b):
    a = a.lower()
    b = b.lower()

    # base check
    if abs(len(a) - len(b)) > 1:
        return False

    a_dict = Counter(a)
    b_dict = Counter(b)

    diff = 0
    for key in a_dict.keys():
        key_value = a_dict[key]

        if key_value > 1:
            if key not in b_dict.keys():
                return False
            else:
                if abs(a_dict[key] - b_dict[key]) > 1:
                    return False
                else:
                    if abs(a_dict[key] - b_dict[key]) == 1:
                        diff += 1
        else:
            if key_value == 1:
                if key not in b_dict.keys():
                    diff += 1
                if key in b_dict.keys():
                    if abs(a_dict[key] - b_dict[key]) > 1:
                        return False
                    elif abs(a_dict[key] - b_dict[key]) == 1:
                        diff += 1
        if diff > 1:
            return False

    return True

if __name__ == "__main__":

    print(oneAway("pale", "ple"))
    print(oneAway("pales", "pale"))
    print(oneAway("pale", "bale"))
    print(oneAway("pale", "bake"))



